### PR TITLE
Allow networking pluging to POST non JSON body Content-Types

### PR DESCRIPTION
**Pull request by GrahO23**
The networking plugin doesn't support POSTs with Content-Types that are not JSON format.

This is an issues as many APIs including the Salesforce token end point ('/services/oauth2/token') don't support JSON Content Type.

As a fix I detect is the Content-Type on a post is "application/x-www-form-urlencoded; charset=UTF-8" and if so add the payload string to a NSDictionary with key  @"x-www-form-urlencoded".

This is then picked up in the POST request in SFRestRequest.m and converted to NSData and POST is completed.

To enable non-JSON Content-Type set the content header to in Javascript

eg

var payload  = "grant_type=refresh_token&client_id=xxx&refresh_token=xxxx"; cordova.require("com.salesforce.plugin.network").sendRequest('https://login.salesforce.com',
        '/services/oauth2/token',
        successCB,
        errorCB,
        "POST",
        payload
        headers
    );

### CODE REVIEW

**Comment by huminzhi** [comment]
ok to test

**Comment by huminzhi** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Code coverage under minimum of 80%</td>
    </tr>
  </tbody>
</table>

## Current coverage for SalesforceAnalytics.framework is `73%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceSDKCore.framework is `51%`
Files changed | - | - 
--- | --- | ---
SFRestRequest.m | `67%` | :warning:
SFRestRequest.m | `67%` | :warning:

---
## Current coverage for CocoaLumberjack.framework is `32%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartStore.framework is `63%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSync.framework is `76%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceHybridSDK.framework is `42%`
Files changed | - | - 
--- | --- | ---
SFNetworkPlugin.m | `72%` | :warning:
SFNetworkPlugin.m | `72%` | :warning:

---
## Current coverage for SalesforceReact.framework is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for RestAPIExplorer.app is `26%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSyncExplorer.app is `19%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSyncExplorerCommon.framework is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for RecentContactsTodayExtension.appex is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for AccountEditor.app is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for NoteSync.app is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSyncExplorerHybrid.app is `0%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceSDKCoreTestApp.app is `30%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartStoreTestApp.app is `61%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSyncTestApp.app is `61%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceHybridSDKTestApp.app is `85%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceAnalyticsTestApp.app is `61%`
✅ *No files affecting coverage found*

---

> Powered by [xcov](https://github.com/nakiostudio/xcov)

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by GrahO23** [comment]
@ huminzhi Thanks 

Do I need to up the coverage for SalesforceSDKCore & SalesforceHybridSDK ?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m [hunk 1]
```diff
@@ -225,8 +225,17 @@ - (void)prepareRequestForSend
         // It's POSTish.  The Network SDK handles content-based requests more or less automatically,
         // but if you want to post a JSON object or other data, you have to manage the contents yourself.
         if (self.action.parameters.parameterStyle != CSFParameterStyleMultipart) {
-            // Standard POST data.  We'll assume we can just send it as JSON.
-            NSData *bodyData = [SFJsonUtils JSONDataRepresentation:self.queryParams];
+            NSData *bodyData = nil;
+            //Check if the POST should be sent as an encoded string rather than JSON
+            NSString* formEncoded = (NSString*)[self.queryParams objectForKey: @"x-www-form-urlencoded" ];
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m]: There are far too many specialized content types to one-off each one in this area of the code.  It's what `[SFRestRequest setCustomRequestBodyString:contentType:]` and its other input analogs were made for.  See my previous comments.

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m [hunk 2]
```diff
@@ -57,18 +58,27 @@ - (void) pgSendRequest:(CDVInvokedUrlCommand *) command
     NSString* path = [argsDict nonNullObjectForKey:kPathArg];
 
     NSDictionary* queryParams = [[NSDictionary alloc] init];
+    NSDictionary<NSString*, NSString*>* headerParams = [argsDict nonNullObjectForKey:kHeaderParams];
```
Comment (wmathurin) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m]: We need the ability to specify any http request headers (custom end point could expect specific headers and the client should have a way to specify them or for instance if-xxx headers for "smarter" end points). This is not the artifact of old functionality.

Content type is one of many "permanent"  http request headers (full list here: https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Request_fields), I don't see a reason to single it out and treat it as a separate parameter.

**Comment by bhariharan** [comment]
Thanks @khawkins. It sounds like we should prioritize this work for a future release, where we can do this in a more generic way.

File: libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m (hunk 2 above)
Comment (khawkins) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m]: > This is not the artifact of old functionality.

Ah, that was me misreading the diff.  It looked like setting the value was new as of this request.  I missed it below.

> I don't see a reason to single it out and treat it as a separate parameter.

My reasoning is that, in a world where we're specifying arbitrary content as input, it makes more sense for the value not to be buried in another collection of properties as an afterthought.  Indeed, that's why Content Type is explicitly called out as a parameter to `SFRestRequest`'s initializer.  We can leave it in a collection, but it feels a bit removed from intention in that use case, to me.

Comment (khawkins) [inline libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m]: Or to think of it another way, IMO that value drives a creation process of the request that includes setting the proper header values, not the other way around.

**Comment by GrahO23** [comment]
@bhariharan Thanks will close the request. Playing with the unstable Mobile SDK with forcejs and that plays really nice on WKWebview.

**Pull request closed without merging**